prow/plank/error.go

```diff
@@ -0,0 +1,31 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plank
+
+import "errors"
+
+var ClusterClientNotExistError *clusterClientNotExistError
+
+type clusterClientNotExistError struct{}
+
+func (ce *clusterClientNotExistError) Error() string {
+	return "Cluster client not exist"
+}
+
+func (ce *clusterClientNotExistError) Unwrap() error {
```

Comment:
I don't think you need an `Unwrap`, you need an `Is` and we need to make sure that the `%w` verb is used with `fmt.Errorf` everywhere (There are linters for that).

A nicer pattern for triggering behavior via errors is to do it as wrapper, because that allows you to easily re-use the triggering behavior part for other use cases (say: Podspec is invalid, that will never succeed either). Something like this: https://github.com/openshift/ci-tools/blob/e253c1cb44c7eed7468086efb3812a02c283c6ee/pkg/controller/util/error.go#L10

Reply:
Looking at the `TerminalError` implementation, agree that it's pretty elegant implementation, but feel like this would exactly be how go1.13 implemented `Is` and `Unwrap` in my imagination, which makes me think it's like re-inventing a very tiny wheel. IMHO using `%w` and `Is` is more `go` way.

For the reuse scenario, I think I can make it more general by renaming the error to be something like `nonRetryableError` so that it can be used by other cases as well.

Reply:
> Looking at the TerminalError implementation, agree that it's pretty elegant implementation, but feel like this would exactly be how go1.13 implemented Is and Unwrap in my imagination, which makes me think it's like re-inventing a very tiny wheel. IMHO using %w and Is is more go way.

What I linked is entirely based on golangs 1.13 errors.Is, not sure I understand the comment?

> For the reuse scenario, I think I can make it more general by renaming the error to be something like nonRetryableError so that it can be used by other cases as well.

That would be great. Also, if we want to keep this working we should probably enforce `%w` usage via a linter

Reply:
> What I linked is entirely based on golangs 1.13 errors.Is, not sure I understand the comment?

Yeah that's right, sorry I overlooked.

 
> > For the reuse scenario, I think I can make it more general by renaming the error to be something like nonRetryableError so that it can be used by other cases as well.
> 
> That would be great. Also, if we want to keep this working we should probably enforce `%w` usage via a linter

I just looked up and couldn't find where golint is configured for this repo, did I overlook something?

Reply:
> I just looked up and couldn't find where golint is configured for this repo, did I overlook something?

We do not have any linting today, it got stuck on @fejta insisting to do that within bazel via repo-infra and no one from the non-googlers knows how do do that: https://github.com/kubernetes/test-infra/pull/19589#discussion_r507889662

Reply:
Gotcha, thanks for all the contexts! Switched to go with TerminalError as wrapper function approach